src/hotspot/share/gc/z/zPhysicalMemory.cpp

```diff
@@ -139,58 +144,202 @@ void ZPhysicalMemoryManager::warn_commit_limits(size_t max) const {
+    const size_t expanded = _backing.commit(start, remaining);
+    if (expanded > 0) {
+      // Successful or partialy successful
+      _committed.free(start, expanded);
+      committed += expanded;
+    }
+  }
+
+  return committed;
 }
 
 size_t ZPhysicalMemoryManager::uncommit(size_t size) {
-  return _backing.uncommit(size);
+  size_t uncommitted = 0;
+
+  // Punch holes in backing memory
+  while (uncommitted < size) {
+    size_t allocated = 0;
+    const size_t remaining = size - uncommitted;
+    const uintptr_t start = _committed.alloc_from_back_at_most(remaining, &allocated);
+    assert(start != UINTPTR_MAX, "Allocation should never fail");
+
+    // Try punch hole
+    const size_t punched = _backing.uncommit(start, allocated);
+    if (punched > 0) {
+      // Successful or partialy successful
+      _uncommitted.free(start, punched);
+      uncommitted += punched;
+    }
+    if (punched < allocated) {
+      // Failed or partialy failed
+      _committed.free(start + punched, allocated - punched);
+      return uncommitted;
+    }
+  }
+
+  return uncommitted;
 }
 
 ZPhysicalMemory ZPhysicalMemoryManager::alloc(size_t size) {
-  return _backing.alloc(size);
+  assert(is_aligned(size, ZGranuleSize), "Invalid size");
+
+  ZPhysicalMemory pmem;
+
+  // Allocate segments
+  for (size_t allocated = 0; allocated < size; allocated += ZGranuleSize) {
+    const uintptr_t start = _committed.alloc_from_front(ZGranuleSize);
+    assert(start != UINTPTR_MAX, "Allocation should never fail");
+    pmem.add_segment(ZPhysicalMemorySegment(start, ZGranuleSize));
+  }
+
+  return pmem;
 }
 
 void ZPhysicalMemoryManager::free(const ZPhysicalMemory& pmem) {
-  _backing.free(pmem);
+  const size_t nsegments = pmem.nsegments();
+
+  // Free segments
+  for (size_t i = 0; i < nsegments; i++) {
+    const ZPhysicalMemorySegment& segment = pmem.segment(i);
+    _committed.free(segment.start(), segment.size());
+  }
+}
+
+void ZPhysicalMemoryManager::pretouch_view(uintptr_t addr, size_t size) const {
+  const size_t page_size = ZLargePages::is_explicit() ? ZGranuleSize : os::vm_page_size();
+  os::pretouch_memory((void*)addr, (void*)(addr + size), page_size);
+}
+
+void ZPhysicalMemoryManager::map_view(const ZPhysicalMemory& pmem, uintptr_t addr) const {
+  const size_t nsegments = pmem.nsegments();
+  size_t size = 0;
+
+  // Map segments
+  for (size_t i = 0; i < nsegments; i++) {
+    const ZPhysicalMemorySegment& segment = pmem.segment(i);
+    _backing.map(addr + size, segment.size(), segment.start());
+    size += segment.size();
+  }
+
+  // Setup NUMA interleaving
+  if (ZNUMA::is_enabled()) {
+    os::numa_make_global((char*)addr, size);
+  }
+
+  // Setup transparent large pages
+  if (ZLargePages::is_transparent()) {
+    os::realign_memory((char*)addr, size, os::large_page_size());
+  }
+}
+
+void ZPhysicalMemoryManager::unmap_view(const ZPhysicalMemory& pmem, uintptr_t addr) const {
+  _backing.unmap(addr, pmem.size());
```

Comment:
This has the similar concern with the last patch. map/unmap is not declared or defined in some platforms, how do we prevent users from compiling with INCLUDE_ZGC and seeing compilation error?